### PR TITLE
ARROW-8392: [Java] Fix overflow related corner cases for vector value comparison

### PR DESCRIPTION
**Pull request by liyafan82**
1. Fix corner cases related to overflow.
2. Provide test cases for the corner cases.

### CODE REVIEW

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 1]
```diff
@@ -212,7 +212,7 @@ public UInt2Comparator() {
     public int compareNotNull(int index1, int index2) {
       char value1 = vector1.get(index1);
       char value2 = vector2.get(index2);
-      return value1 - value2;
+      return (value1 & 0xffff) - (value2 & 0xffff);
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java:215]: it looks like there is Character.compare?

**Review by emkornfield — CHANGES_REQUESTED** [review]
I think this also needs to be done for byte and short?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 1 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: > it looks like there is Character.compare?

Thanks for your kind reminder.
The implementation of Character#compare is as follows:

```
    public static int compare(char x, char y) {
        return x - y;
    }
```
In effect, the chars x and y are first promoted to int32 (automatically, according to Java specification), and then a minus operation is performed.
So the values x and y are actually compared as signed integers, which is not what we want here.

**Comment by liyafan82** [comment]
> I think this also needs to be done for byte and short?

According to the Java specification, byte and short integers are automatically promoted to int32, when performing arithmetic operations, so maybe these types do not have the problem of integer overflow.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 1 above)
Comment (emkornfield) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Could you add documentation inline about this so I don't forget the next time we look at the code?

Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java:219]: Sure. Thank you for the good suggestion.

**Comment by emkornfield** [comment]
+1 thank you!

**Pull request closed without merging**